File: plugins/agentic-sdlc/src/agentic_sdlc/steps/ralph_loop_step.py
"""Ralph loop step executor."""

from __future__ import annotations

from typing import TYPE_CHECKING

from agentic_sdlc.console import extract_summary
from agentic_sdlc.progress import WorkflowStatus, update_step_completed, update_step_failed
from agentic_sdlc.ralph_loop import (
    build_ralph_system_message,
    create_ralph_state,
    deactivate_ralph_state,
    detect_completion_promise,
    update_ralph_iteration,
)
from agentic_sdlc.runner import run_claude
from agentic_sdlc.steps.base import StepContext, StepExecutor, StepResult

if TYPE_CHECKING:
    from agentic_sdlc.console import ConsoleOutput
    from agentic_sdlc.logging.logger import WorkflowLogger
    from agentic_sdlc.parser import StepDefinition
    from agentic_sdlc.progress import WorkflowProgress


class RalphLoopStepExecutor(StepExecutor):
    """Executes a Ralph Wiggum loop step.

    The Ralph loop runs the same prompt iteratively until:
    - Claude outputs a completion JSON with the expected promise
    - Maximum iterations is reached
    """

    def execute(
        self,
        step: StepDefinition,
        progress: WorkflowProgress,
        context: StepContext,
        logger: WorkflowLogger,
        console: ConsoleOutput,
    ) -> StepResult:
        """Execute a Ralph Wiggum loop step."""
        prompt_template = step.prompt or ""
        completion_promise = step.completion_promise or "COMPLETE"
        template_context = context.build_template_context()

        if context.renderer.has_variables(completion_promise):
            completion_promise = context.renderer.render_string(completion_promise, template_context)

        # Handle max_iterations which may be a template string or integer
        max_iterations_raw = step.max_iterations
        if isinstance(max_iterations_raw, str) and context.renderer.has_variables(max_iterations_raw):
            max_iterations = int(context.renderer.render_string(max_iterations_raw, template_context))
        else:
            max_iterations = int(max_iterations_raw) if max_iterations_raw else 5

        timeout = (step.step_timeout_minutes or 30) * 60
        bypass_permissions = context.workflow_settings.bypass_permissions if context.workflow_settings else False
        allowed_tools = context.workflow_settings.required_tools if context.workflow_settings else None

        # Always enable streaming when console is provided (handles both BASE and ALL modes)
        print_output = True

        logger.info(step.name, f"Starting Ralph loop (max {max_iterations} iterations)")
        console.info(f"Ralph loop starting (max {max_iterations} iterations)")

        all_outputs: list[str] = []

        for iteration in range(1, max_iterations + 1):
            logger.info(step.name, f"Ralph loop iteration {iteration}/{max_iterations}")

            # Render prompt with iteration context for each iteration
            iteration_context = {
                **template_context,
                "iteration": iteration,
                "max_iterations": max_iterations,
            }
            if context.renderer.has_variables(prompt_template):
                prompt = context.renderer.render_string(prompt_template, iteration_context)
            else:
                prompt = prompt_template

            # Create ralph state on first iteration (after we have the rendered prompt)
            if iteration == 1:
                _ralph_state = create_ralph_state(
                    workflow_id=progress.workflow_id,
                    step_name=step.name,
                    prompt=prompt,
                    max_iterations=max_iterations,
                    completion_promise=completion_promise,
                    repo_root=context.repo_root,
                )

            system_message = build_ralph_system_message(iteration, max_iterations, completion_promise)
            full_prompt = f"{system_message}{prompt}"

            result = run_claude(
                prompt=full_prompt,
                cwd=context.repo_root,
                model=context.resolve_model(step.model),
                timeout=timeout,
                print_output=print_output,
                skip_permissions=bypass_permissions,
                allowed_tools=allowed_tools,
                console=console,
                workflow_id=context.workflow_id,
            )

            if not result.success:
                error_summary = extract_summary(result.stderr) if result.stderr else "Unknown error"
                logger.warning(step.name, f"Iteration {iteration} failed: {result.stderr}")
                console.ralph_iteration(step.name, iteration, max_iterations, f"Failed: {error_summary}")
                if iteration < max_iterations:
                    update_ralph_iteration(progress.workflow_id, step.name, context.repo_root)
                    continue
                else:
                    deactivate_ralph_state(progress.workflow_id, step.name, context.repo_root)
                    error_msg = f"Ralph loop failed after {max_iterations} iterations"
                    console.step_failed(step.name, error_msg)
                    update_step_failed(progress, step.name, error_msg)
                    progress.status = WorkflowStatus.FAILED.value
                    return StepResult(success=False, error=error_msg)

            all_outputs.append(result.stdout)

            # Print iteration summary
            iteration_summary = extract_summary(result.stdout)
            console.ralph_iteration(step.name, iteration, max_iterations, iteration_summary)

            completion_result = detect_completion_promise(result.stdout, completion_promise)

            # Check for failure signal first
            if completion_result.is_failed:
                error_msg = f"Ralph loop failed: {completion_result.failure_reason}"
                logger.error(step.name, error_msg)
                deactivate_ralph_state(progress.workflow_id, step.name, context.repo_root)
                combined_output = "\n\n---\n\n".join(all_outputs)
                console.step_failed(step.name, error_msg)
                update_step_failed(progress, step.name, error_msg)
                progress.status = WorkflowStatus.FAILED.value
                return StepResult(success=False, error=error_msg, full_output=combined_output)

            if completion_result.is_complete and completion_result.promise_matched:
                logger.info(step.name, f"Ralph loop completed at iteration {iteration}")
                deactivate_ralph_state(progress.workflow_id, step.name, context.repo_root)
                combined_output = "\n\n---\n\n".join(all_outputs)
                output_summary = f"Completed in {iteration} iterations"
                update_step_completed(progress, step.name, output_summary, combined_output)
                console.ralph_complete(step.name, iteration, max_iterations)
                return StepResult(success=True, output_summary=output_summary, full_output=combined_output)

            if iteration < max_iterations:
                update_ralph_iteration(progress.workflow_id, step.name, context.repo_root)

        logger.warning(step.name, f"Ralph loop reached max iterations ({max_iterations}) without completion")
        deactivate_ralph_state(progress.workflow_id, step.name, context.repo_root)
        combined_output = "\n\n---\n\n".join(all_outputs)
        output_summary = f"Max iterations ({max_iterations}) reached without completion promise"
        update_step_completed(progress, step.name, output_summary, combined_output)
        console.ralph_max_iterations(step.name, max_iterations)
        return StepResult(success=True, output_summary=output_summary, full_output=combined_output)

File: src/steps/ralph-loop-step.ts
/** Ralph loop step executor. */

import type { ConsoleOutput } from "../console.js";
import { extractSummary } from "../console.js";
import type { WorkflowLogger } from "../logging/logger.js";
import { getOutputDir } from "../paths.js";
import { WORKFLOW_STATUS, updateStepCompleted, updateStepFailed } from "../progress.js";
import {
	buildRalphSystemMessage,
	createRalphState,
	deactivateRalphState,
	detectCompletionPromise,
	loadRalphState,
	updateRalphIteration,
} from "../ralph-loop.js";
import { runRuntime } from "../runtimes/process-runner.js";
import type { StepDefinition, WorkflowProgress } from "../types.js";
import {
	type StepContext,
	StepExecutor,
	type StepResult,
	buildTemplateContext,
	resolveModel,
} from "./base.js";

export class RalphLoopStepExecutor extends StepExecutor {
	async execute(
		step: StepDefinition,
		progress: WorkflowProgress,
		context: StepContext,
		logger: WorkflowLogger,
		console: ConsoleOutput,
	): Promise<StepResult> {
		const promptTemplate = step.prompt ?? "";
		const templateContext = buildTemplateContext(context);

		let completionPromise = step.completionPromise ?? "COMPLETE";
		if (context.renderer.hasVariables(completionPromise)) {
			completionPromise = context.renderer.renderString(completionPromise, templateContext);
		}

		// Handle max_iterations which may be a template string or integer
		const maxIterationsRaw = step.maxIterations;
		let maxIterations: number;
		if (typeof maxIterationsRaw === "string" && context.renderer.hasVariables(maxIterationsRaw)) {
			maxIterations = Number.parseInt(
				context.renderer.renderString(maxIterationsRaw, templateContext),
				10,
			);
		} else {
			maxIterations = maxIterationsRaw ? Number.parseInt(String(maxIterationsRaw), 10) : 5;
		}

		const timeout = (step.stepTimeoutMinutes ?? 30) * 60;
		const bypassPermissions = context.workflowSettings?.bypassPermissions ?? false;
		const allowedTools = context.workflowSettings?.requiredTools?.length
			? context.workflowSettings.requiredTools
			: null;

		// Always enable streaming when console is provided
		const printOutput = true;

		logger.info(step.name, `Starting Ralph loop (max ${maxIterations} iterations)`);
		console.info(`Ralph loop starting (max ${maxIterations} iterations)`);

		const allOutputs: string[] = [];

		const outputDir =
			context.outputDir ?? getOutputDir(progress.workflowId, context.config, context.repoRoot);

		// Check for existing ralph state to resume from
		const existingState = loadRalphState(progress.workflowId, step.name, outputDir);
		let startIteration: number;
		if (existingState?.active) {
			startIteration = existingState.iteration;
			logger.info(step.name, `Resuming Ralph loop from iteration ${startIteration}`);
			console.info(`Resuming Ralph loop from iteration ${startIteration}`);
		} else {
			startIteration = 1;
		}

		for (let iteration = startIteration; iteration <= maxIterations; iteration++) {
			logger.info(step.name, `Ralph loop iteration ${iteration}/${maxIterations}`);

			// Render prompt with iteration context
			const iterationContext = {
				...templateContext,
				iteration,
				max_iterations: maxIterations,
			};
			let prompt: string;
			if (context.renderer.hasVariables(promptTemplate)) {
				prompt = context.renderer.renderString(promptTemplate, iterationContext);
			} else {
				prompt = promptTemplate;
			}

			// Create ralph state on first iteration
			if (iteration === startIteration && !existingState?.active) {
				createRalphState(
					progress.workflowId,
					step.name,
					prompt,
					maxIterations,
					completionPromise,
					outputDir,
				);
			}

			const systemMessage = buildRalphSystemMessage(iteration, maxIterations, completionPromise);
			const fullPrompt = `${systemMessage}${prompt}`;

			// Display iteration header BEFORE running runtime
			console.ralphIterationStart(step.name, iteration, maxIterations);

			const result = await runRuntime(context.runtimeAdapter, {
				prompt: fullPrompt,
				cwd: context.repoRoot,
				model: resolveModel(context, step.model),
				timeout,
				printOutput,
				skipPermissions: bypassPermissions,
				allowedTools,
				console,
				workflowId: context.workflowId,
				outputDir: context.outputDir,
			});

			if (!result.success) {
				const errorSummary = result.stderr ? extractSummary(result.stderr) : "Unknown error";
				logger.warning(step.name, `Iteration ${iteration} failed: ${result.stderr}`);
				console.ralphIteration(step.name, iteration, maxIterations, `Failed: ${errorSummary}`);
				if (iteration < maxIterations) {
					updateRalphIteration(progress.workflowId, step.name, outputDir);
					continue;
				}
				deactivateRalphState(progress.workflowId, step.name, outputDir);
				const errorMsg = `Ralph loop failed after ${maxIterations} iterations`;
				console.stepFailed(step.name, errorMsg);
				updateStepFailed(progress, step.name, errorMsg);
				progress.status = WORKFLOW_STATUS.FAILED;
				return { success: false, error: errorMsg };
			}

			allOutputs.push(result.stdout);

			// Print iteration summary
			const iterationSummary = extractSummary(result.stdout);
			console.ralphIteration(step.name, iteration, maxIterations, iterationSummary);

			const completionResult = detectCompletionPromise(result.stdout, completionPromise);

			// Check for failure signal first
			if (completionResult.isFailed) {
				const errorMsg = `Ralph loop failed: ${completionResult.failureReason}`;
				logger.error(step.name, errorMsg);
				deactivateRalphState(progress.workflowId, step.name, outputDir);
				const combinedOutput = allOutputs.join("\n\n---\n\n");
				console.stepFailed(step.name, errorMsg);
				updateStepFailed(progress, step.name, errorMsg);
				progress.status = WORKFLOW_STATUS.FAILED;
				return { success: false, error: errorMsg, fullOutput: combinedOutput };
			}

			if (completionResult.isComplete && completionResult.promiseMatched) {
				logger.info(step.name, `Ralph loop completed at iteration ${iteration}`);
				deactivateRalphState(progress.workflowId, step.name, outputDir);
				const combinedOutput = allOutputs.join("\n\n---\n\n");
				const outputSummary = `Completed in ${iteration} iterations`;
				updateStepCompleted(progress, step.name, outputSummary, combinedOutput);
				console.ralphComplete(step.name, iteration, maxIterations);
				return { success: true, outputSummary, fullOutput: combinedOutput };
			}

			if (iteration < maxIterations) {
				updateRalphIteration(progress.workflowId, step.name, outputDir);
			}
		}

		logger.warning(
			step.name,
			`Ralph loop reached max iterations (${maxIterations}) without completion`,
		);
		deactivateRalphState(progress.workflowId, step.name, outputDir);
		const combinedOutput = allOutputs.join("\n\n---\n\n");
		const outputSummary = `Max iterations (${maxIterations}) reached without completion promise`;
		updateStepCompleted(progress, step.name, outputSummary, combinedOutput);
		console.ralphMaxIterations(step.name, maxIterations);
		return { success: true, outputSummary, fullOutput: combinedOutput };
	}
}
